### PR TITLE
Volume observer pr3

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Replace the current `VolumeDataFile` class with a new `VolumeData` class that is a subfile rather than a separate H5 file.
- Add a couple tags needed for IO

*Note:* Does not depend on anything else currently open.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/IO/H5/File.hpp [hunk 1]
```diff
@@ -241,6 +248,33 @@ ObjectType& H5File<Access_t>::insert(const std::string& path,
...
+  current_object_ = nullptr;
+  // C++17: structured bindings
+  auto exists_group_name = check_if_object_exists<ObjectType>(path);
+  if (std::get<0>(exists_group_name)) {
+    hid_t group_id = std::get<1>(exists_group_name).id();
+    current_object_ = std::make_unique<ObjectType>(
+        std::get<0>(exists_group_name),
+        std::move(std::get<1>(exists_group_name)), group_id,
+        std::move(std::get<2>(exists_group_name)), std::forward<Args>(args)...);
+    return dynamic_cast<ObjectType&>(*current_object_);
```
Comment (wthrowe) [inline src/IO/H5/File.hpp]: Why not `convert_to_derived`?

File: tests/Unit/IO/Test_H5.cpp [hunk 2]
```diff
@@ -218,8 +218,12 @@ SPECTRE_TEST_CASE("Unit.IO.H5.Dat", "[Unit][IO][H5]") {
   std::vector<std::string> legend{"Time", "Error L2", "Error L1", "Error"};
 
   h5::H5File<h5::AccessType::ReadWrite> my_file(h5_file_name);
+  {
+    auto& error_file =
+        my_file.insert<h5::Dat>("/L2_errors", legend, version_number);
+  }
   auto& error_file =
-      my_file.insert<h5::Dat>("/L2_errors", legend, version_number);
+      my_file.try_insert<h5::Dat>("/L2_errors", legend, version_number);
```
Comment (wthrowe) [inline tests/Unit/IO/Test_H5.cpp]: Pass different arguments to verify that it doesn't overwrite.

File: src/IO/H5/VolumeData.cpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+  // Write extents.
+  const auto& extents = extents_and_tensors.extents;
+  if (not contains_attribute(spatial_group.id(), "", "extents")) {
+    h5::write_to_attribute(spatial_group.id(), "extents", extents);
+  }
+
+  // Write the connectivity.
+  if (not h5::contains_dataset_or_group(spatial_group.id(), "",
+                                        "connectivity")) {
+    const std::vector<int> connectivity = [&extents]() {
```
Comment (wthrowe) [inline src/IO/H5/VolumeData.cpp]: noexcept

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/IO/H5/VolumeData.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+                       ? name
+                       : name + extension())
+                : name + extension()),
+      version_(version),
+      volume_file_root_group_(group_.id(), name_, h5::AccessType::ReadWrite) {
+  if (exists) {
+    const Version open_version(true, detail::OpenGroup{},
+                               volume_file_root_group_.id(), "version");
+    version_ = open_version.get_version();
+    const Header header(true, detail::OpenGroup{}, volume_file_root_group_.id(),
```
Comment (kidder) [inline src/IO/H5/VolumeData.cpp]: add error if input version doesn't match version_, (and maybe a comment about backwards compatibility when we have more than one version)

File: src/IO/H5/File.hpp [hunk 5]
```diff
@@ -241,6 +248,33 @@ ObjectType& H5File<Access_t>::insert(const std::string& path,
...
+  }
+
+  hid_t group_id = std::get<1>(exists_group_name).id();
```
Comment (kidder) [inline src/IO/H5/File.hpp]: what is different between the two cases?

**Review by kidder — CHANGES_REQUESTED** [review]

File: tests/Unit/IO/Test_H5.cpp [hunk 6]
```diff
@@ -223,8 +218,15 @@ SPECTRE_TEST_CASE("Unit.IO.H5.Dat", "[Unit][IO][H5]") {
   std::vector<std::string> legend{"Time", "Error L2", "Error L1", "Error"};
 
   h5::H5File<h5::AccessType::ReadWrite> my_file(h5_file_name);
+  {
+    auto& error_file =
+        my_file.insert<h5::Dat>("/L2_errors", legend, version_number);
+  }
+  // pass bad values to make sure it aren't overridden.
```
Comment (kidder) [inline tests/Unit/IO/Test_H5.cpp]: they aren't or it isn't

Comment (nilsdeppe) [reply tests/Unit/IO/Test_H5.cpp]: Okay, rebased and fixed (hopefully :P )

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**